### PR TITLE
cdc_acm: fix interrupt handling, API registration, and propgate errors.

### PR DESCRIPTION
**Pull request by nzmichaelh**
As per uart_pipe.c, irq_is_pending should only return true if any
*unmasked* interrupts are pending.

The device_get_binding() code only checks devices that have an API.
Set the API at compile time so that other devices can refer to the CDC
ACM driver.

usb_write() may return EBUSY if the endpoint is full.  Propagate this
for future use.

Tested on the arduino_zero.

Signed-off-by: Michael Hope <mlhx@google.com>

### CODE REVIEW

**Comment by carlescufi** [comment]
CC @IvanSanchez @sundar-subramaniyan

File: subsys/usb/class/cdc_acm.c [hunk 1]
```diff
@@ -422,9 +420,11 @@ static int cdc_acm_fifo_fill(struct device *dev,
 	len = len > sizeof(u32_t) ? sizeof(u32_t) : len;
 #endif
 
-	usb_write(CONFIG_CDC_ACM_IN_EP_ADDR, tx_data, len, &bytes_written);
-
-	return bytes_written;
+	err = usb_write(CONFIG_CDC_ACM_IN_EP_ADDR, tx_data, len, &wrote);
+	if (err != 0) {
+		return err;
```
Comment (IvanSanchez) [inline subsys/usb/class/cdc_acm.c]: This seems to contradict the docstring `@return Number of bytes sent.`. What is the return value for this function call?

**Review by IvanSanchez — CHANGES_REQUESTED** [review]

File: subsys/usb/class/cdc_acm.c [hunk 2]
```diff
@@ -422,9 +420,11 @@ static int cdc_acm_fifo_fill(struct device *dev,
 	len = len > sizeof(u32_t) ? sizeof(u32_t) : len;
```
Comment (sundar-subramaniyan) [inline subsys/usb/class/cdc_acm.c]: Not relevant to the change made below. The Quark limitation can be removed here as it is taken care in the usb_dc_dw.c driver now. @finikorg Could you please confirm?

Comment (finikorg) [reply subsys/usb/class/cdc_acm.c]: @sundar-subramaniyan I thought it should but we have one issue I am currently investigating #5678. Let's keep this for a while.

File: subsys/usb/class/cdc_acm.c (hunk 1 above)
Comment (carlescufi) [reply subsys/usb/class/cdc_acm.c]: I think this should be `if (err < 0)`, since return values are typically -ERRCODE on failure

Comment (nzmichaelh) [reply subsys/usb/class/cdc_acm.c]: Hi @IvanSanchez @carlescufi stepping back, there's two types of return value in Zephyr:

- errors encoded as an int, where 0 means success
- result with an optional error.  Encoded as an int where >= 0 is the result and < 0 is the error.

If usb_write() succeeds then it sets wrote and returns 0.  If it fails, it writes nothing and returns an error code.  So testing against 0 is fine.

There's no effective difference between err < 0 and err != 0 it's just the latter is slightly more correct :)

Comment (carlescufi) [reply subsys/usb/class/cdc_acm.c]: Agreed, there is no fundamental difference here

**Review by carlescufi — APPROVED** [review]
This looks fine to me. @finikorg any additional remarks?

**Review by finikorg — APPROVED** [review]
Please next time split code to logical commits and use prefix
"usb: cdc_acm: Fix ...."

**Pull request merged**